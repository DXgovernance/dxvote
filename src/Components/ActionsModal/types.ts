import { DecodedAction } from 'old-components/Guilds/ActionsBuilder/types';

export interface ActionModalProps {
  isOpen: boolean;
  setIsOpen: (isOpen: boolean) => void;
  onAddAction: (action: DecodedAction) => void;
}
