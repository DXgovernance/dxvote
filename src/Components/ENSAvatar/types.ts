export interface ENSAvatarProps {
  address?: string;
  size?: number;
}
