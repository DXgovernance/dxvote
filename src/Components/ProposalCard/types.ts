import { ProposalStatusProps } from 'Components/ProposalStatus/types';
import { Proposal, ENSAvatar } from '../Types';

export interface ProposalCardProps {
  proposal?: Proposal;
  votes?: number[];
  ensAvatar?: ENSAvatar;
  href?: string;
  statusProps?: ProposalStatusProps;
}
