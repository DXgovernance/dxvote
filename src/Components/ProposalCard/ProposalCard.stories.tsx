import { ComponentStory, ComponentMeta } from '@storybook/react';
import ProposalCard from 'Components/ProposalCard/ProposalCard';
import { ProposalCardProps } from 'Components/ProposalCard/types';
import { proposalStatusMock } from 'Components/ProposalStatus/fixtures';
import { ensAvatarMock, proposalMock } from '../Fixtures';

export default {
  title: 'Proposals/ProposalCard',
  component: ProposalCard,
  argTypes: {
    href: {
      description: 'URL to open on click',
    },
    votes: {
      description: 'Array of vote percentages for each option',
    },
  },
} as ComponentMeta<typeof ProposalCard>;

const proposalCardArgs: ProposalCardProps = {
  proposal: proposalMock,
  ensAvatar: ensAvatarMock,
  href: '/',
  votes: [11, 1],
  statusProps: proposalStatusMock,
};

const Template: ComponentStory<typeof ProposalCard> = args => (
  <ProposalCard {...args} />
);

const ManyTemplate: ComponentStory<typeof ProposalCard> = args => (
  <div>
    <Template {...args} />
    <Template {...args} />
    <Template />
  </div>
);

export const Simple = Template.bind({});
Simple.args = proposalCardArgs;

export const Loading = Template.bind({});
Loading.args = {};

export const Group = ManyTemplate.bind({});
Group.args = proposalCardArgs;
