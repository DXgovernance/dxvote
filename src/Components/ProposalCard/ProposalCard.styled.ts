import styled from 'styled-components';
import { Box } from 'Components/Primitives/Layout';
import { Heading } from 'old-components/Guilds/common/Typography';

// TODO: base these components on a generic Card component
export const CardWrapper = styled(Box)`
  border: 1px solid ${({ theme }) => theme.colors.muted};
  border-radius: ${({ theme }) => theme.radii.curved};
  margin-bottom: 1rem;
  padding: 1rem;
  color: ${({ theme }) => theme.colors.proposalText.lightGrey};
  &:hover {
    border-color: ${({ theme }) => theme.colors.border.hover};
    color: ${({ theme }) => theme.colors.text};
  }
`;

export const CardHeader = styled(Box)`
  display: flex;
  flex-direction: row;
  justify-content: space-between;
`;

export const CardContent = styled(Box)`
  margin: 1rem 0;
`;

export const CardFooter = styled(Box)`
  display: flex;
  flex-direction: row;
  justify-content: space-between;
  align-items: flex-end;
  @media only screen and (max-width: 524px) {
    flex-direction: column;
    align-items: flex-start;
  }
`;

export const CardTitle = styled(Heading)`
  font-size: 1rem;
  font-weight: 700;
  color: ${({ theme }) => theme.colors.text};
  @media only screen and (min-width: 768px) {
    font-size: 1.25rem;
  }
`;

export const IconDetailWrapper = styled(Box)`
  display: flex;
  flex-direction: row;
  align-items: center;
  flex: 1;
`;

export const Detail = styled(Box)`
  font-size: 0.95rem;
  font-weight: 600;
  margin-left: 0.5rem;
`;

export const ActionsWrapper = styled(Box)`
  display: flex;
  flex: 1;
  margin-right: 24px;
  position: relative;
  overflow-x: hidden;
  &::before {
    content: '';
    position: absolute;
    top: 0;
    left: 0;
    right: 0;
    width: 100%;
    height: 100%;
    background: ${({ theme }) =>
      `linear-gradient(to right, transparent 89%, ${theme.colors.background} 100%)`};
  }
  & > div {
    margin: 4px 2px;
  }
  @media only screen and (max-width: 524px) {
    flex-wrap: wrap;
    border-bottom: 1px solid ${({ theme }) => theme.colors.muted};
    margin-bottom: 0.5rem;
    padding-bottom: 0.5rem;
    & > div:nth-child(n + 3) {
      display: none;
    }
    &::before {
      content: none;
    }
  }
`;

export const BorderedIconDetailWrapper = styled(IconDetailWrapper)`
  border: 1px solid ${({ theme }) => theme.colors.border.initial};
  border-radius: 1rem;
  padding: 0.25rem 0.8rem;
  flex: none;
  display: flex;
  width: fit-content;
`;

export const NotFoundActionWrapper = styled.div`
  display: flex;
  padding: 4px;
  border: ${({ theme }) => `1px solid ${theme.colors.red}`};
  border-radius: 30px;
`;

export const Icon = styled.img<{
  spaceLeft?: boolean;
  spaceRight?: boolean;
  bordered: boolean;
}>`
  width: 1.5rem;
  height: 1.5rem;
  display: flex;
  justify-content: center;
  align-items: center;

  ${props => props.spaceLeft && `margin-left: 0.5rem;`}
  ${props => props.spaceRight && `margin-right: 0.5rem;`}

  ${props =>
    props.bordered &&
    `
    border: 1px solid #000;
    border-radius: 50%;
  `}
`;

export const VoteInfoWrapper = styled(Box)`
  min-width: unset;
`;
