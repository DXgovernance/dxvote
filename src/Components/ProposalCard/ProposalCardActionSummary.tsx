import { Loading } from 'Components/Primitives/Loading';
import { getInfoLineView } from 'old-components/Guilds/ActionsBuilder/SupportedActions';
import UndecodableCallInfoLine from 'old-components/Guilds/ActionsBuilder/UndecodableCalls/UndecodableCallsInfoLine';
import { useTypedParams } from 'Modules/Guilds/Hooks/useTypedParams';
import { useFilteredProposalActions } from 'hooks/Guilds/guild/useFilteredProposalActions';
import { Proposal } from 'Components/Types';
import {
  ActionsWrapper,
  BorderedIconDetailWrapper,
  NotFoundActionWrapper,
} from 'Components/ProposalCard/ProposalCard.styled';

interface ProposalCardActionSummaryProps {
  proposal?: Proposal;
}

const ProposalCardActionSummary: React.FC<ProposalCardActionSummaryProps> = ({
  proposal,
}) => {
  const { guildId } = useTypedParams();
  const actions = useFilteredProposalActions(guildId, proposal?.id);
  const isLoading = !proposal;
  if (isLoading) {
    return (
      <Loading
        style={{ margin: 0 }}
        loading
        text
        skeletonProps={{ width: '200px' }}
      />
    );
  }
  return (
    <ActionsWrapper>
      {actions?.map(action => {
        if (!action) return null;
        const InfoLine = getInfoLineView(action?.decodedCall?.callType);

        return !!InfoLine ? (
          <BorderedIconDetailWrapper>
            <InfoLine
              decodedCall={action?.decodedCall}
              approveSpendTokens={action?.approval}
              compact
            />
          </BorderedIconDetailWrapper>
        ) : (
          <NotFoundActionWrapper>
            <UndecodableCallInfoLine />
          </NotFoundActionWrapper>
        );
      })}
    </ActionsWrapper>
  );
};

export default ProposalCardActionSummary;
