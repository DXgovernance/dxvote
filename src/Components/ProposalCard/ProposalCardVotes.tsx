import { Fragment } from 'react';
import { FiCircle } from 'react-icons/fi';
import { Loading } from 'Components/Primitives/Loading';
import {
  Detail,
  BorderedIconDetailWrapper,
  Icon,
  VoteInfoWrapper,
} from 'components/ProposalCard/ProposalCard.styled';

interface ProposalCardVotesProps {
  isLoading?: boolean;
  votes?: number[];
}

const ProposalCardVotes: React.FC<ProposalCardVotesProps> = ({
  isLoading,
  votes,
}) => {
  if (isLoading) {
    return (
      <Loading
        style={{ margin: 0 }}
        loading
        text
        skeletonProps={{ width: '200px' }}
      />
    );
  }
  return (
    <VoteInfoWrapper>
      <BorderedIconDetailWrapper>
        {votes
          .sort((a, b) => b - a)
          .map((vote, i) => {
            if (i < 3 && !(i === votes.length - 1)) {
              return (
                <Fragment>
                  <Detail>{vote}%</Detail>
                  <Icon as="div" spaceLeft spaceRight>
                    <FiCircle />
                  </Icon>
                </Fragment>
              );
            } else {
              return <Detail>{vote}%</Detail>;
            }
          })}
      </BorderedIconDetailWrapper>
    </VoteInfoWrapper>
  );
};

export default ProposalCardVotes;
