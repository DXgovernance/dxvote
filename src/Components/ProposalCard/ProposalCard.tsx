import ProposalCardActionSummary from 'Components/ProposalCard/ProposalCardActionSummary';
import ProposalCardVotes from 'Components/ProposalCard/ProposalCardVotes';
import { ProposalCardProps } from 'Components/ProposalCard/types';
import Avatar from 'old-components/Guilds/Avatar';
import ProposalStatus from 'Components/ProposalStatus/ProposalStatus';
import { Loading } from 'Components/Primitives/Loading';
import UnstyledLink from 'old-components/Guilds/common/UnstyledLink';
import 'react-loading-skeleton/dist/skeleton.css';
import { shortenAddress } from 'utils';
import {
  CardWrapper,
  CardHeader,
  IconDetailWrapper,
  Detail,
  CardContent,
  CardTitle,
  CardFooter,
} from 'Components/ProposalCard/ProposalCard.styled';

const ProposalCard: React.FC<ProposalCardProps> = ({
  proposal,
  votes,
  ensAvatar,
  href,
  statusProps,
}) => {
  return (
    <UnstyledLink to={href || '#'}>
      <CardWrapper>
        <CardHeader>
          <IconDetailWrapper>
            {proposal?.creator ? (
              <Avatar
                src={ensAvatar.imageUrl}
                defaultSeed={proposal.creator}
                size={24}
              />
            ) : (
              <Loading
                style={{ margin: 0 }}
                loading
                text
                skeletonProps={{ circle: true, width: '24px', height: '24px' }}
              />
            )}
            <Detail>
              {ensAvatar?.ensName ||
                (proposal?.creator ? (
                  shortenAddress(proposal.creator)
                ) : (
                  <Loading style={{ margin: 0 }} loading text />
                ))}
            </Detail>
          </IconDetailWrapper>
          <ProposalStatus {...statusProps} />
        </CardHeader>
        <CardContent>
          <CardTitle size={2}>
            <strong>
              {proposal?.title || (
                <Loading style={{ margin: 0 }} loading text />
              )}
            </strong>
          </CardTitle>
        </CardContent>
        <CardFooter>
          <ProposalCardActionSummary isLoading={!proposal} />
          <ProposalCardVotes isLoading={!proposal} votes={votes} />
        </CardFooter>
      </CardWrapper>
    </UnstyledLink>
  );
};

export default ProposalCard;
