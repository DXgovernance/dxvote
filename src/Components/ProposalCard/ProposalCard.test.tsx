import ProposalCard from './ProposalCard';
import { render } from '../../utils/tests';
import '@testing-library/jest-dom';
import { ProposalCardProps } from 'Components/ProposalCard/types';
import moment from 'moment';
import { ProposalState } from 'Components/Types';

const validProps: ProposalCardProps = {
  proposal: {
    id: '42',
    creator: '0x0b17cf48420400e1d71f8231d4a8e43b3566bb5b',
    startTime: moment(),
    endTime: moment(),
    to: [''],
    data: [''],
    value: [],
    totalActions: null,
    title: 'test',
    contentHash: '0x0',
    state: ProposalState.Active,
    totalVotes: [],
  },
  votes: [10, 20],
  ensAvatar: { ensName: 'rossdev.eth', imageUrl: '' },
  href: 'testUrl',
  statusProps: {
    timeDetail: 'Time',
    status: ProposalState.Active,
  },
};

const invalidProps: ProposalCardProps = {
  proposal: null,
  votes: [],
  ensAvatar: null,
  href: null,
  statusProps: {
    timeDetail: null,
    status: null,
  },
};

test('ProposalCard Renders properly with data', async () => {
  const { container } = render(<ProposalCard {...validProps} />);

  expect(container).toMatchSnapshot();
});

test('ProposalCard loading', async () => {
  const { container } = render(<ProposalCard {...invalidProps} />);

  expect(container).toMatchSnapshot();
});
