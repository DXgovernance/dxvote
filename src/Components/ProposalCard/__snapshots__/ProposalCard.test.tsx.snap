// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProposalCard ProposalCard Renders properly with data 1`] = `
<div>
   
  <a
    class="sc-bBXrwG fDAvrC"
    href="/testUrl"
  >
    <div
      class="sc-dlfnuX sc-iBPTik jmIxJI hyNmzl"
    >
      <div
        class="sc-dlfnuX sc-fubCzh jmIxJI cXVqBt"
      >
        <div
          class="sc-dlfnuX sc-iqHYmW jmIxJI eOgsPb"
        >
          <div
            class="paper"
            style="border-radius: 50px; display: inline-block; margin: 0px; overflow: hidden; padding: 0px; background-color: rgb(251, 24, 107); height: 24px; width: 24px;"
          >
            <svg
              height="24"
              width="24"
              x="0"
              xmlns="http://www.w3.org/2000/svg"
              y="0"
            >
              <rect
                fill="#F93301"
                height="24"
                rx="0"
                ry="0"
                transform="translate(-3.3540020288113044 -0.24569355652725883) rotate(184.4 12 12)"
                width="24"
                x="0"
                y="0"
              />
              <rect
                fill="#F5B800"
                height="24"
                rx="0"
                ry="0"
                transform="translate(-14.879522566232806 -1.4785377082910314) rotate(308.7 12 12)"
                width="24"
                x="0"
                y="0"
              />
              <rect
                fill="#03555E"
                height="24"
                rx="0"
                ry="0"
                transform="translate(21.86674072242258 -1.2091297142613349) rotate(454.0 12 12)"
                width="24"
                x="0"
                y="0"
              />
              <rect
                fill="#C81456"
                height="24"
                rx="0"
                ry="0"
                transform="translate(-22.985221694835005 -17.58014285250981) rotate(346.4 12 12)"
                width="24"
                x="0"
                y="0"
              />
            </svg>
          </div>
          <div
            class="sc-dlfnuX sc-crrszt jmIxJI xiXTo"
          >
            rossdev.eth
          </div>
        </div>
        <div
          class="sc-hHfuMS cMWNCN"
        >
          <div
            class="sc-dmlqKv hjvFsS"
          >
            <div
              class="sc-dlfnuX sc-fKFxtB jmIxJI eZHbZU"
            >
              <span
                title="May 9th, 2022 - 8:00 am"
              >
                Time
              </span>
            </div>
            <div
              class="sc-dlfnuX sc-kfzBvY jmIxJI jgJnFX"
            >
               
              Active
            </div>
          </div>
        </div>
      </div>
      <div
        class="sc-dlfnuX sc-pGacB jmIxJI jwWloK"
      >
        <h2
          class="sc-gKseQn sc-kEjbQP fMfHRo eUCyLV"
          size="2"
        >
          <strong>
            test
          </strong>
        </h2>
      </div>
      <div
        class="sc-dlfnuX sc-jrAFXE jmIxJI gIuaJs"
      >
        <div
          class="sc-dlfnuX sc-fodVek jmIxJI enmJQB"
        >
          <div
            class="sc-dlfnuX sc-iqHYmW sc-bqyKOL jmIxJI eOgsPb hdYgXU"
          >
            <div
              class="sc-dlfnuX sc-crrszt jmIxJI xiXTo"
            >
              20
              %
            </div>
            <div
              class="sc-hBEYId bLtNkZ"
            >
              <svg
                fill="none"
                height="1em"
                stroke="currentColor"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="2"
                viewBox="0 0 24 24"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <circle
                  cx="12"
                  cy="12"
                  r="10"
                />
              </svg>
            </div>
            <div
              class="sc-dlfnuX sc-crrszt jmIxJI xiXTo"
            >
              10
              %
            </div>
          </div>
        </div>
      </div>
    </div>
  </a>
   
</div>
`;

exports[`ProposalCard ProposalCard loading 1`] = `
<div>
   
  <a
    class="sc-bBXrwG fDAvrC"
    href="/"
  >
    <div
      class="sc-dlfnuX sc-iBPTik jmIxJI hyNmzl"
    >
      <div
        class="sc-dlfnuX sc-fubCzh jmIxJI cXVqBt"
      >
        <div
          class="sc-dlfnuX sc-iqHYmW jmIxJI eOgsPb"
        >
          <div
            class="sc-bdfBQB"
            style="margin: 0px;"
          >
            <span
              aria-busy="true"
              aria-live="polite"
            >
              <span
                class="react-loading-skeleton"
                style="width: 24px; height: 24px; border-radius: 50%; --base-color: #333; --highlight-color: #555;"
              >
                ‌
              </span>
              <br />
            </span>
          </div>
          <div
            class="sc-dlfnuX sc-crrszt jmIxJI xiXTo"
          >
            <div
              class="sc-bdfBQB"
              style="margin: 0px;"
            >
              <span
                aria-busy="true"
                aria-live="polite"
              >
                <span
                  class="react-loading-skeleton"
                  style="width: 100px; height: 16px; border-radius: 50px; --base-color: #333; --highlight-color: #555;"
                >
                  ‌
                </span>
                <br />
              </span>
            </div>
          </div>
        </div>
        <div
          class="sc-hHfuMS cMWNCN"
        >
          <div
            class="sc-dmlqKv hjvFsS"
          >
            <div
              class="sc-dlfnuX sc-fKFxtB jmIxJI eZHbZU"
            >
              <div
                class="sc-bdfBQB"
              >
                <span
                  aria-busy="true"
                  aria-live="polite"
                >
                  <span
                    class="react-loading-skeleton"
                    style="width: 50px; --base-color: #333; --highlight-color: #555;"
                  >
                    ‌
                  </span>
                  <br />
                </span>
              </div>
            </div>
            <div
              class="sc-bdfBQB"
            >
              <span
                aria-busy="true"
                aria-live="polite"
              >
                <span
                  class="react-loading-skeleton"
                  style="width: 50px; --base-color: #333; --highlight-color: #555;"
                >
                  ‌
                </span>
                <br />
              </span>
            </div>
          </div>
        </div>
      </div>
      <div
        class="sc-dlfnuX sc-pGacB jmIxJI jwWloK"
      >
        <h2
          class="sc-gKseQn sc-kEjbQP fMfHRo eUCyLV"
          size="2"
        >
          <strong>
            <div
              class="sc-bdfBQB"
              style="margin: 0px;"
            >
              <span
                aria-busy="true"
                aria-live="polite"
              >
                <span
                  class="react-loading-skeleton"
                  style="width: 100px; height: 16px; border-radius: 50px; --base-color: #333; --highlight-color: #555;"
                >
                  ‌
                </span>
                <br />
              </span>
            </div>
          </strong>
        </h2>
      </div>
      <div
        class="sc-dlfnuX sc-jrAFXE jmIxJI gIuaJs"
      >
        <div
          class="sc-bdfBQB"
          style="margin: 0px;"
        >
          <span
            aria-busy="true"
            aria-live="polite"
          >
            <span
              class="react-loading-skeleton"
              style="width: 200px; --base-color: #333; --highlight-color: #555;"
            >
              ‌
            </span>
            <br />
          </span>
        </div>
      </div>
    </div>
  </a>
   
</div>
`;
